src/main/java/com/minecolonies/entity/ai/basic/AbstractEntityAIBasic.java

```diff
@@ -745,21 +756,94 @@ private boolean dumpOneMoreSlot(@NotNull Predicate<ItemStack> keepIt)
+                  return amountToKeep == 0;
               }
-              worker.getInventoryCitizen().decrStackSize(i, stack.stackSize - returnStack.stackSize);
-              //Check that we are not inserting
-              // into a full inventory.
+              worker.getInventoryCitizen().decrStackSize(i, stack.stackSize - returnStack.stackSize - amountToKeep);
+              //Check that we are not inserting into a full inventory.
               return stack.stackSize != returnStack.stackSize;
           });
     }
 
     /**
+     * Handle the cases when X items should be kept.
+     * @param keptX already kept items.
+     * @param toKeep to keep items.
+     * @param tempStorage item to analyze.
+     * @return null if should be kept entirely, else itemStack with amount which should be dumped.
+     */
+    private static ItemStack handleKeepX(Map<ItemStorage, Integer> keptX, Map<ItemStorage, Integer> toKeep, ItemStorage tempStorage)
+    {
+        int dump;
+
+        if(keptX.get(tempStorage) == null)
+        {
+            if (toKeep.get(tempStorage) >= tempStorage.getAmount())
+            {
+                keptX.put(tempStorage, tempStorage.getAmount());
+                return null;
+            }
+
+            dump = tempStorage.getAmount() - toKeep.get(tempStorage);
+            keptX.put(tempStorage, toKeep.get(tempStorage));
+        }
+        else
+        {
+            final int amountKept = keptX.remove(tempStorage);
+            if (toKeep.get(tempStorage) >= (tempStorage.getAmount() + amountKept))
+            {
+                keptX.put(tempStorage, tempStorage.getAmount() + amountKept);
+                return null;
+            }
+            keptX.put(tempStorage, toKeep.get(tempStorage));
+            dump = tempStorage.getAmount() + amountKept - toKeep.get(tempStorage);
+        }
+
+        //Create tempStack with the amount of items that should be dumped.
+        return new ItemStack(tempStorage.getItem(), dump, tempStorage.getDamageValue());
+    }
+
+    /**
+     * Checks if enough items have been kept already.
+     * @param kept kept items.
+     * @param keep items to keep.
+     * @param stack stack to analyse.
+     * @return true if kept enough already.
+     */
+    private static boolean keptEnough(Map<ItemStorage, Integer> kept, Map<ItemStorage, Integer> keep, ItemStack stack)
+    {
+        for(ItemStorage tempStack: keep.keySet())
+        {
+            if(tempStack.getItem() == stack.getItem() && tempStack.getDamageValue() != stack.getItemDamage())
+            {
+                keep.put(new ItemStorage(stack.getItem(), stack.getItemDamage(), 0, tempStack.ignoreDamageValue()), keep.get(tempStack));
+                break;
+            }
+        }
+        final ItemStorage tempStorage = new ItemStorage(stack.getItem(), stack.getItemDamage(), 0, false);
+        return keep.get(tempStorage) == null || (kept.get(tempStorage) != null && kept.get(tempStorage) >= keep.get(tempStorage));
```

Comment:
This return is fairly confusing. we should simplify it or at least add a comment explaining stuff